pkg/crdhandler/handler.go

```diff
@@ -0,0 +1,173 @@
+		h.logger.Errorf(ctx, err, "unable to parse admission review request")
+		metrics.InvalidRequests.WithLabelValues("mutating").Inc()
+		writer.WriteHeader(http.StatusBadRequest)
+		return
+	}
+
+	response, err := doConversion(review.Request, h.converter)
+	if err != nil {
+		var uid types.UID
+		if review.Request != nil {
+			uid = review.Request.UID
+		}
+		h.writeResponse(ctx, request, writer, h.writeErrorResponse(uid, err))
+		return
+	}
+
+	h.logger.Debugf(request.Context(), "converted")
+	metrics.SuccessfulRequests.WithLabelValues("converting").Inc()
+
+	h.writeResponse(ctx, request, writer, response)
+}
+
+func New(config Config) (Handler, error) {
+	serializers := map[mediaType]runtime.Serializer{
+		{"application", "json"}: json.NewSerializerWithOptions(json.DefaultMetaFactory, config.Scheme, config.Scheme, json.SerializerOptions{
+			Yaml:   false,
+			Pretty: false,
+			Strict: false,
+		}),
+		{"application", "yaml"}: json.NewSerializerWithOptions(json.DefaultMetaFactory, config.Scheme, config.Scheme, json.SerializerOptions{
+			Yaml:   true,
+			Pretty: false,
+			Strict: false,
+		}),
+	}
+
+	return Handler{
+		logger:      config.Logger,
+		converter:   config.Converter,
+		serializers: serializers,
+	}, nil
+}
+
+// doConversion converts the requested object given the conversion function and returns a conversion response.
+// failures will be reported as Reason in the conversion response.
+func doConversion(convertRequest *v1.ConversionRequest, converter Converter) (*v1.ConversionResponse, error) {
+	if convertRequest == nil {
+		return nil, errors.New("invalid request")
+	}
+
+	var convertedObjects []runtime.RawExtension
+	for _, obj := range convertRequest.Objects {
+		cr := unstructured.Unstructured{}
+		if err := cr.UnmarshalJSON(obj.Raw); err != nil {
+			return nil, microerror.Mask(err)
+		}
+		convertedCR, err := converter.Convert(&cr, convertRequest.DesiredAPIVersion)
+		if err != nil {
+			return nil, microerror.Mask(err)
+		}
+		convertedCR.SetAPIVersion(convertRequest.DesiredAPIVersion)
+		convertedObjects = append(convertedObjects, runtime.RawExtension{Object: convertedCR})
+	}
+
+	return &v1.ConversionResponse{
+		ConvertedObjects: convertedObjects,
+		UID:              convertRequest.UID,
+		Result: metav1.Status{
+			Status: metav1.StatusSuccess,
+		},
+	}, nil
+}
+
+func (h Handler) writeResponse(ctx context.Context, request *http.Request, writer http.ResponseWriter, response *v1.ConversionResponse) {
+	accept := request.Header.Get("Accept")
+	outSerializer, err := h.getOutputSerializer(accept)
+	if err != nil {
+		h.logger.Errorf(ctx, err, fmt.Sprintf("invalid Accept header `%s`", accept))
+		metrics.InvalidRequests.WithLabelValues("mutating").Inc()
+		writer.WriteHeader(http.StatusBadRequest)
+		return
+	}
+
+	err = outSerializer.Encode(&v1.ConversionReview{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "ConversionReview",
+			APIVersion: "apiextensions.k8s.io/v1",
```

Comment:
I wasn't sure about this `APIVersion`. Could someone confirm that this is correct?